common/transport/amqp/src/amqp.ts

```diff
@@ -132,6 +88,185 @@ export class Amqp {
         forever: false
       }
     }, amqp10.Policy.EventHub));
+
+    this._fsm = new machina.Fsm({
+      initialState: 'disconnected',
+      states: {
+        disconnected: {
+          connect: (connectCallback) => {
+            this._fsm.transition('connecting');
+            let connectError = null;
+            const connectErrorHandler = (err) => {
+              connectError = err;
+            };
+
+            const amqpErrorHandler = (err) => {
+              debug('amqp10 error: ' + err.toString());
+              if (this._disconnectHandler) {
+                this._disconnectHandler(err);
+              }
+            };
+
+            this._amqp.on(_amqpClientError, connectErrorHandler);
+            this._amqp.connect(this.uri)
+              .then((result) => {
+                debug('AMQP transport connected.');
+                this._amqp.on(_amqpClientError, amqpErrorHandler);
+                this._amqp.removeListener(_amqpClientError, connectErrorHandler);
+                this._fsm.transition('connected');
+                /*Codes_SRS_NODE_COMMON_AMQP_16_002: [The `connect` method shall establish a connection with the IoT hub instance and if given as argument call the `done` callback with a null error object in the case of success and a `results.Connected` object.]*/
+                this._safeCallback(connectCallback, null, new results.Connected(result));
+                return null;
+              })
+              .catch((err) => {
+                this._amqp.removeListener(_amqpClientError, connectErrorHandler);
+                this._fsm.transition('disconnected');
+                /*Codes_SRS_NODE_COMMON_AMQP_16_003: [The `connect` method shall call the `done` callback if the connection fails.] */
+                this._safeCallback(connectCallback, connectError || err);
+              });
+          },
+          disconnect: (callback) => callback(null, new results.Disconnected()),
+          attachSenderLink: (endpoint, linkOptions, callback) => {
+            this._fsm.handle('connect', (err) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('attachSenderLink', endpoint, linkOptions, callback);
+              }
+            });
```

Comment:
I've put a lot more thought to this today, trying to find a better design, because honestly this was designed lazily at first. turns out lazy might actually be good, here's what I have in mind to explain why: 

the main challenge here is that we want to call the action-specific callback on connection error or action error, but only on action success if connection succeeds: Leaving it here allows to group the error and success case of the connect action in a very readable manner, whereas using state transitions and queues split the call of the callback in multiple places (in _onEnter of the disconnected state for connection errors and in _onEnter of the connected state for connection success, where we'd handle the actual action). It also requires complicated queue logic (significantly more complicated than for link objects, because  we'd have many different types of actions in the same queue, whereas links have only one type of action).

whether we call "handle('connect', ...)" or "transition('connecting', ...)" and use the _onEnter handler is a different question: in that case, I feel both are mostly equivalent and valid ways to write that code. What tips the scale for me is that I like the idea that the transition to connecting/connected is all done in the body of the connect action handler, and I like how easy it is to read (looks more imperative as opposed to transitioning to the "connecting" state and assuming the reader knows that the connection happens just because we transitioned).

Lastly, there's the question of what happens when we start a bunch of actions in a row (eg. 2 attachSenderLink and a send): should we fail all of them if the first one fails to connect, or should we try to connect for each action that was requested by the caller? Since the retry logic implemented by the caller may be different for each type of action though, i'm not sure it makes sense to "group fail" everything if the first action's call to connect fails. To rephrase, intuitively, I think each action "deserves a shot" at connecting.
To the argument that we have a different pattern for link attach and connect, I think it's acceptable because connecting is a lot more error prone than attaching links: it depends on more external factors (if the connection is up and authenticated, attaching the link should succeed, attaching doesn't depend on wifi quality, etc: the connection does) so it also feels ok treating connect and attach differently.

It's all work in progress though and those arguments might not stand the test of reimplementing the client layer. I guess we'll see, but feel free to challenge those assumptions: it's a complex piece of code.


---
In reply to: [129133265](https://github.com/Azure/azure-iot-sdk-node/pull/55#discussion_r129133265) [](ancestors = 129133265,127617317)